src/libraries/System.Data.Common/src/System/Data/Common/ObjectStorage.cs

```diff
@@ -236,50 +236,49 @@ public override void Set(int recordNo, object value)
             }
             else
             {
-                Type valType = value.GetType();
-                if (_dataType == typeof(Guid) && valType == typeof(string))
+                if (_dataType == typeof(Guid) && value is string)
                 {
                     _values[recordNo] = new Guid((string)value);
                 }
                 else if (_dataType == typeof(byte[]))
                 {
-                    if (valType == typeof(bool))
+                    if (value is bool)
```

Comment:
nit: can use a switch expression

Reply:
This may hit the performance issue logged in #47920. I actually have a branch to improve this, but it's not expandable-proof so I haven't push it.

It's also interesting to see if a following `unbox.any` changes the codegen of `isinst`.

Reply:
In the case of a switch expression, the JIT is not able to eliminate the calls to `CORINFO_HELP_UNBOX`.

Reply:
I'm surprised that `CORINFO_HELP_UNBOX` is ever evolved after #87241.

Reply:
> I'm surprised that `CORINFO_HELP_UNBOX` is ever evolved after #87241.

The `isinst` + `unbox.any` optimization appears in .NET 7.